Adding skeleton device->host service worker.

benvanik opened the pull request:
The worker is assigned to nodes nearby the device they are serving and mimics an HSA/AQL queue. The intent is that HSA-specific tooling can intercept our packets but I'm not sure there's any _good_ tooling that does that we care about. Since the difference between an AQL queue and what we'd do is small I decided to stick to the standard here. In the future we can evaluate whether it makes more sense to have our own packet format (instead of agent dispatches) or our own ringbuffers for common operations like releasing resources or signaling semaphores from the host.

benvanik committed:
Adding skeleton device->host service worker.
The worker is assigned to nodes nearby the device they are serving and
mimics an HSA/AQL queue. The intent is that HSA-specific tooling can
intercept our packets but I'm not sure there's any _good_ tooling that
does that we care about. Since the difference between an AQL queue
and what we'd do is small I decided to stick to the standard here.

ScottTodd commented on runtime/src/iree/hal/drivers/amdgpu/device/support/queue.h:
```diff
@@ -130,6 +130,66 @@ typedef struct IREE_AMDGPU_ALIGNAS(64) iree_amd_queue_t {
+  // Pointer to the ringbuffer read dispatch ID atomic.
+  iree_amdgpu_scoped_atomic_uint64_t* read_dispatch_id;
+
+  // Backing queue that may live in host or device memory.
+  iree_amd_queue_t* queue;
+} iree_amd_cached_queue_t;
```
Are some parts of the driver using `iree_amd_queue_t` while others now use `iree_amd_cached_queue_t`? When should one be used over the other? Is it worth having multiple implementations, or could they be merged?

benvanik replied:
The (iree_)amd_queue_t is the hardware/firmware queue and not owned by us (nor allocated by us). iree_amd_cached_queue_t does what it says above by caching some hot fields so we don't have to indirect through the hardware queue: if the queue lives in host memory and we are running on device doing even a `queue->size` query requires reading uncached host memory - and that happens on every packet operation. When writing a packet the `base_address` is required before any writes can happen and having to go to uncached host memory to fetch it and _then_ perform pointer arithmetic to find the packet offset and _then_ write the packet is too much latency.

ScottTodd commented on runtime/src/iree/hal/drivers/amdgpu/device/host_client.h:
```diff
@@ -0,0 +1,115 @@
+// Represents the host runtime thread that is managing host interrupts.
+// One or more schedulers may share a single host queue. An host calls that need
+// to identify the scheduler or scheduler-related resources must pass those as
+// arguments.
```
nit: typo in here
```suggestion
// Represents the host runtime thread that is managing host interrupts.
// One or more schedulers may share a single host queue. Any host calls that need
// to identify the scheduler or scheduler-related resources must pass those as
// arguments.
```

ScottTodd commented on runtime/src/iree/hal/drivers/amdgpu/device/host_client.h:
```diff
@@ -0,0 +1,115 @@
+  // Host softqueue processing device requests. May be servicing requests from
+  // multiple device agents. Cached inline in device memory but references a
+  // ringbuffer in host memory.
+  iree_amd_cached_queue_t service_queue;
```
Typo? "softqueue"?

benvanik replied:
ah soft-queue

ScottTodd commented on runtime/src/iree/hal/drivers/amdgpu/device/host_client.h:
```diff
@@ -0,0 +1,115 @@
+// Enqueues a unidirection host agent packet ("post").
+// Since this is device->host only operation this uses an acquire scope only
+// from the agent and releases to the entire system so the host agent can
+// observe changes. The completion signal is optional and may be
```
I'm confused by the grammar here

```suggestion
// Since this is a device->host only operation this only uses an acquire scope
// from the agent and releases to the entire system so the host agent can
// observe changes. The completion signal is optional and may be
```

benvanik replied:
me too :)

ScottTodd commented on runtime/src/iree/hal/drivers/amdgpu/host_service.c:
```diff
@@ -0,0 +1,606 @@
+      IREE_LIBHSA(libhsa), IREE_ARRAYSIZE(packet->dep_signal),
+      (hsa_signal_t*)packet->dep_signal, conds, values, UINT64_MAX,
+      HSA_WAIT_STATE_BLOCKED,
+      /*satisfying_values=*/NULL);
```
What is the significance of `5` here? Is that meant to line up with some other part of the code? If we had `// LINT.IFCHANGE` style tooling, would we use it here?

benvanik replied:
It's the field count in the HSA spec for the hsa_barrier_and_packet_t - it'll never change as it's baked into hardware. 5 was chosen because all packets are 64 bytes and that's as many as they could fit in one alongside the other packet metadata. Added a static assert for readability.

ScottTodd commented on runtime/src/iree/hal/drivers/amdgpu/host_service.c:
```diff
@@ -0,0 +1,606 @@
+  // Release each resource. Some entries may be NULL.
+  for (iree_host_size_t i = 0; i < 4; ++i) {
+    iree_hal_resource_release(resources[i]);
+  }
```
More `5` here. I guess I'm wondering if one is added (somehow) from 5 to 6, how would we know all the other locations in the code to update?

benvanik replied:
4, not 5! ;P
This is also baked into the spec as part of hsa_agent_dispatch_packet_t (which has 4 args). I could probably craft a static assert for readability, but it'll never change (it'd have to be a new packet type/format). To make the connection clearer I'll pass in a count and let the compiler try to figure out the unrolling.

ScottTodd commented on runtime/src/iree/hal/drivers/amdgpu/host_service.c:
```diff
@@ -0,0 +1,606 @@
+  IREE_IGNORE_ERROR(iree_hal_amdgpu_host_service_barrier(service, libhsa));
+
+  IREE_TRACE_ZONE_END(z0);
+  return 0;
```
Should this ever return nonzero? Maybe if `iree_hal_amdgpu_host_service_fail` is called? There is a status above.

benvanik replied:
We don't use the platform features for propagating errors as their representation/quality differs across platforms (pthreads returns a void* on join, but we only call join during deinit where there's nothing we could possible do with that info). The only way we propagate errors is via our own failure status field.

benvanik committed:
Making AMDGPU tests reuse the loaded libhsa/topology.
ROCR is (mostly) stateless once initialized and there's no need to
reload the library and requery the whole topology for every test case.

benvanik committed:
Adding host_service_test, fixing issues, and applying review feedback.

benvanik committed:
Renaming iree_hal_amdgpu_device_host_client_call_t for consistency.
